### ISSUE TITLE
Launching containers not recognized by Docker daemon

### ISSUE THREAD

**Issue by mark-church** [moby/moby#21632]
This example test done was done on an AWS 40 cpu box in a customer's testing environment and no other load was running except this test program. The spindocker.sh that performed the test and the corresponding output is below.

First noticed the issue when launching 500 containers though marathon in this 40 cpu node. Each docker container creation is a Mesos task. Marathon will hit the default task launch timeout of 5 mins after creating few hundred containers and suspend the deploy for too many failures. The failure was confusing because the tasks that failed, tried to inspect the creation of the container for 5 mins before it gave up. However, the node showed all those containers to be created but stopped. Why did mesos fail to find those containers?

So, after reading through mesos docker containerizer and docker executor code and ruling out mesos issues, wrote the test program mimicking the situation and taking mesos out of the picture.

**Output of `docker info`:**

```
root@ip-xxx-249:~# docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 1.10.3-cs2
Storage Driver: devicemapper
 Pool Name: docker-202:1-525873-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: ext4
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 1.246 GB
 Data Space Total: 107.4 GB
 Data Space Available: 106.1 GB
 Metadata Space Used: 26.76 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.121 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.77 (2012-10-15)
Execution Driver: native-0.2
Logging Driver: json-file
Plugins:
 Volume: local
 Network: bridge null host
Kernel Version: 4.2.8-040208-generic
Operating System: Ubuntu 14.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 40
Total Memory: 157.4 GiB
Name: ip-xxx-249
ID: PMQQ:QMBD:5TTT:I7ZS:DWPR:U34W:PIQV:UGWJ:4R6T:DNQ7:OIXH:JZAG
Debug mode (server): true
 File Descriptors: 12
 Goroutines: 65
 System Time: 2016-03-28T23:37:02.144079146Z
 EventsListeners: 1
 Init SHA1: 37dbf45df08ae9208f38ba8f4eb7d837b5dbe06e
 Init Path: /usr/lib/docker/dockerinit
 Docker Root Dir: /var/lib/docker
WARNING: No swap limit support
```

**spindocker.sh**

```
#!/bin/bash

# nifty function copied from stackoverflow
kill_descendant_processes() {
    local pid="$1"
    local and_self="${2:-false}"
    if children="$(pgrep -P "$pid")"; then
        for child in $children; do
            kill_descendant_processes "$child" true
        done
    fi
    if [[ "$and_self" == true ]]; then
        echo "killing $pid"
        kill  -9 "$pid"
    fi
}

echo "Starting program at $(date)"
container_prefix=$1
# Spin 500 subprocess in parallel in background simulating 500 tasks sent to mesos slave/agent
# each of these subprocess then spawns another child subprocess in background that does docker run while it monitors the creation of the container
# if container creation not detected after a while, the subprocess kills all its child and its grandchild. So, both the child subprocess and docker process will get killed
# this closely mimics the mesos docker containerizer and mesos docker executor model
# program terminates when all subprocess exits
for count in {1..500}
do
(
   (
      docker pull nginx:latest
      echo "Launching container ${container_prefix}_$count on $(date)"
      docker run -p 80 --name ${container_prefix}_$count nginx
   ) &

   COUNT=0
   while true; do
     sleep 1
     docker inspect ${container_prefix}_$count > /dev/null
     if [ $? -eq 0 ]; then
          break
     fi
     let COUNT=COUNT+1
     if [ $COUNT -gt 210 ]; then
         echo "##################### Aborting tries of ${container_prefix}_$count at $(date)"
         docker stop ${container_prefix}_$count
         kill_descendant_processes $BASHPID
         break
     fi
   done


)&
done
wait
echo "all work done at $(date)"

```

**Results and log file**

```

docker status before the test program launch

root@ip-xxx-249:~# docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 1.10.3-cs2
Storage Driver: devicemapper
 Pool Name: docker-202:1-525873-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: ext4
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 1.246 GB
 Data Space Total: 107.4 GB
 Data Space Available: 106.1 GB
 Metadata Space Used: 26.76 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.121 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.77 (2012-10-15)
Execution Driver: native-0.2
Logging Driver: json-file
Plugins:
 Volume: local
 Network: bridge null host
Kernel Version: 4.2.8-040208-generic
Operating System: Ubuntu 14.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 40
Total Memory: 157.4 GiB
Name: ip-xxx-249
ID: PMQQ:QMBD:5TTT:I7ZS:DWPR:U34W:PIQV:UGWJ:4R6T:DNQ7:OIXH:JZAG
Debug mode (server): true
 File Descriptors: 12
 Goroutines: 65
 System Time: 2016-03-28T23:37:02.144079146Z
 EventsListeners: 1
 Init SHA1: 37dbf45df08ae9208f38ba8f4eb7d837b5dbe06e
 Init Path: /usr/lib/docker/dockerinit
 Docker Root Dir: /var/lib/docker
WARNING: No swap limit support


root@ip-xxx-249:~# docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES


Essentially no containers.



Program is now launched and finished.

root@ip-xxx-249:~# time ./spindocker.sh spintest &> outfile

real    3m58.506s
user    75m8.908s
sys     15m11.124s


Lets now look at the key issue and logs.

Below is the start and exit time of the program mentioning the timestamps from the output file.

Logs from outfile for start and end time
----------------------------------------

"Starting program at Mon Mar 28 23:39:56 UTC 2016"

"all work done at Mon Mar 28 23:43:54 UTC 2016"


As the program was running, started checking the progress in a different terminal

Other terminal checking progress
--------------------------------

root@ip-xxx-249:~# date;docker info
Mon Mar 28 23:42:35 UTC 2016
Containers: 252
 Running: 250
 Paused: 0
 Stopped: 2


Greping for containers in created state. As you see below at any point there may be few containers at create state and then vanishes as it moves into running state...

root@ip-xxx-249:~# docker ps -a | grep Create
8829cdbc204c        nginx               "nginx -g 'daemon off"   2 minutes ago       Created                                                  spintest_194
fed3ffe808ef        nginx               "nginx -g 'daemon off"   2 minutes ago       Created                                                  spintest_324
root@ip-xxx-249:~# docker ps -a | grep Create
root@ip-xxx-249:~# docker ps -a | grep Create

root@ip-xxx-249:~# docker ps -a | grep Create
ffac95faebfc        nginx               "nginx -g 'daemon off"   3 minutes ago       Created                                                  spintest_312
9952ae2b9ad7        nginx               "nginx -g 'daemon off"   3 minutes ago       Created                                                  spintest_315
root@ip-xxx-249:~# docker ps -a | grep Create
c75e648f33bb        nginx               "nginx -g 'daemon off"   3 minutes ago       Created                                                  spintest_208



Still progressing..

root@ip-xxx-249:~# date;docker info
Mon Mar 28 23:43:52 UTC 2016
Containers: 373
 Running: 371
 Paused: 0
 Stopped: 2


Program terminated at this point 2 seconds after the above at 23:43:54 UTC 2016. Now the fun/weird behavior will be noticed.

See how the stopped container counts start increasing after program termination...


root@ip-xxx-249:~# date;docker info
Mon Mar 28 23:44:07 UTC 2016
Containers: 404
 Running: 373
 Paused: 0
 Stopped: 31



root@ip-xxx-249:~# date;docker info
Mon Mar 28 23:44:22 UTC 2016
Containers: 437
 Running: 373
 Paused: 0
 Stopped: 64


root@ip-xxx-249:~# date;docker info
Mon Mar 28 23:47:44 UTC 2016
Containers: 500
 Running: 373
 Paused: 0
 Stopped: 127

So, basically after the program terminated, the daemon created 127 stop containers in create state.

So, lets pick any such container (spintest_17 conatiner name)


In docker event log, the container creation event is after the program termination time at 23:44. See below snippet

2016-03-28T23:44:04.014754407Z container create bb4f30e433aa962897b4d8df158ec26f81ac246eb168b64df7a3b8e8b90b1050 (image=nginx, name=spintest_17)

Now docker inspect at spintest_17 as below reports creation time at 23:40 (reflecting probably when daemon got the create request for this container but confusing because creation time should reflect the final creation as emitted from the event log.)

root@ip-xxx-249:~# docker inspect spintest_17
[
    {
        "Id": "bb4f30e433aa962897b4d8df158ec26f81ac246eb168b64df7a3b8e8b90b1050",
        "Created": "2016-03-28T23:40:04.300536345Z",
        "Path": "nginx",
        "Args": [
            "-g",
            "daemon off;"
        ],
        "State": {
            "Status": "created",
            "Running": false,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 0,
            "ExitCode": 0,
            "Error": "",
            "StartedAt": "0001-01-01T00:00:00Z",
            "FinishedAt": "0001-01-01T00:00:00Z"
        },
        "Image": "sha256:af4b3d7d5401624ed3a747dc20f88e2b5e92e0ee9954aab8f1b5724d7edeca5e",
        "ResolvConfPath": "",
        "HostnamePath": "",
        "HostsPath": "",
        "LogPath": "",
        "Name": "/spintest_17",
        "RestartCount": 0,
        "Driver": "devicemapper",
        "MountLabel": "",
        "ProcessLabel": "",
        "AppArmorProfile": "",
        "ExecIDs": null,
        "HostConfig": {
            "Binds": null,
            "ContainerIDFile": "",
            "LogConfig": {
                "Type": "json-file",
                "Config": {}
            },
            "NetworkMode": "default",
            "PortBindings": {
                "80/tcp": [
                    {
                        "HostIp": "",
                        "HostPort": ""
                    }
                ]
            },
            "RestartPolicy": {
                "Name": "no",
                "MaximumRetryCount": 0
            },
            "VolumeDriver": "",
            "VolumesFrom": null,
            "CapAdd": null,
            "CapDrop": null,
            "Dns": [],
            "DnsOptions": [],
            "DnsSearch": [],
            "ExtraHosts": null,
            "GroupAdd": null,
            "IpcMode": "",
            "Links": null,
            "OomScoreAdj": 0,
            "PidMode": "",
            "Privileged": false,
            "PublishAllPorts": false,
            "ReadonlyRootfs": false,
            "SecurityOpt": null,
            "UTSMode": "",
            "ShmSize": 67108864,
            "ConsoleSize": [
                0,
                0
            ],
            "Isolation": "",
            "CpuShares": 0,
            "CgroupParent": "",
            "BlkioWeight": 0,
            "BlkioWeightDevice": null,
            "BlkioDeviceReadBps": null,
            "BlkioDeviceWriteBps": null,
            "BlkioDeviceReadIOps": null,
            "BlkioDeviceWriteIOps": null,
            "CpuPeriod": 0,
            "CpuQuota": 0,
            "CpusetCpus": "",
            "CpusetMems": "",
            "Devices": [],
            "KernelMemory": 0,
            "Memory": 0,
            "MemoryReservation": 0,
            "MemorySwap": 0,
            "MemorySwappiness": -1,
            "OomKillDisable": false,
            "PidsLimit": 0,
            "Ulimits": null
        },
        "GraphDriver": {
            "Name": "devicemapper",
            "Data": {
                "DeviceId": "32300",
                "DeviceName": "docker-202:1-525873-1a9eae265a8ab718bd9fd5f14dce6b9aab0b38e4768c16fd66ad59a59884ad74",
                "DeviceSize": "10737418240"
            }
        },
        "Mounts": [],
        "Config": {
            "Hostname": "bb4f30e433aa",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": true,
            "AttachStderr": true,
            "ExposedPorts": {
                "443/tcp": {},
                "80/tcp": {}
            },
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "NGINX_VERSION=1.9.12-1~jessie"
            ],
            "Cmd": [
                "nginx",
                "-g",
                "daemon off;"
            ],
            "Image": "nginx",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": {},
            "StopSignal": "SIGTERM"
        },
        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": null,
            "SandboxKey": "",
            "SecondaryIPAddresses": null,
            "SecondaryIPv6Addresses": null,
            "EndpointID": "",
            "Gateway": "",
            "GlobalIPv6Address": "",
            "GlobalIPv6PrefixLen": 0,
            "IPAddress": "",
            "IPPrefixLen": 0,
            "IPv6Gateway": "",
            "MacAddress": "",
            "Networks": {
                "bridge": {
                    "IPAMConfig": null,
                    "Links": null,
                    "Aliases": null,
                    "NetworkID": "",
                    "EndpointID": "",
                    "Gateway": "",
                    "IPAddress": "",
                    "IPPrefixLen": 0,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": ""
                }
            }
        }
    }
]


The daemon log below is not very interesting.

[[37mDEBU^[[0m[3920] POST /v1.22/containers/create?name=spintest_17

^[[37mDEBU^[[0m[3920] form data: {"AttachStderr":true,"AttachStdin":false,"AttachStdout":true,"Cmd":null,"Domainname":"","Entrypoint":null,"Env":[],"ExposedPorts":{"80/tcp":{}},"HostConfig":{"Binds":null,"BlkioDeviceReadBps":null,"BlkioDeviceReadIOps":null,"BlkioDeviceWriteBps":null,"BlkioDeviceWriteIOps":null,"BlkioWeight":0,"BlkioWeightDevice":null,"CapAdd":null,"CapDrop":null,"CgroupParent":"","ConsoleSize":[0,0],"ContainerIDFile":"","CpuPeriod":0,"CpuQuota":0,"CpuShares":0,"CpusetCpus":"","CpusetMems":"","Devices":[],"Dns":[],"DnsOptions":[],"DnsSearch":[],"ExtraHosts":null,"GroupAdd":null,"IpcMode":"","Isolation":"","KernelMemory":0,"Links":null,"LogConfig":{"Config":{},"Type":""},"Memory":0,"MemoryReservation":0,"MemorySwap":0,"MemorySwappiness":-1,"NetworkMode":"default","OomKillDisable":false,"OomScoreAdj":0,"PidMode":"","PidsLimit":0,"PortBindings":{"80/tcp":[{"HostIp":"","HostPort":""}]},"Privileged":false,"PublishAllPorts":false,"ReadonlyRootfs":false,"RestartPolicy":{"MaximumRetryCount":0,"Name":"no"},"SecurityOpt":null,"ShmSize":0,"UTSMode":"","Ulimits":null,"VolumeDriver":"","VolumesFrom":null},"Hostname":"","Image":"nginx","Labels":{},"NetworkingConfig":{"EndpointsConfig":{}},"OnBuild":null,"OpenStdin":false,"StdinOnce":false,"StopSignal":"SIGTERM","Tty":false,"User":"","Volumes":{},"WorkingDir":""}


Now lets look at some interesting snippets from the program output file

Logs from outfile for spintest_17
---------------------------------

Launching container spintest_17 on Mon Mar 28 23:40:03 UTC 2016
(Above matches the docker inspect create timestamp of 23:40:04)

##################### Aborting tries of spintest_17 at Mon Mar 28 23:43:51 UTC 2016
(program gives up after configured number of attempts, 210 attempts sleeping one second in between, hence at little over 3 and half mins)

Failed to stop container (spintest_17): Error response from daemon: No such container: spintest_17

(docker stop fails proving container not recognized by daemon)
```

SO, in summary, for this particular container, the program tried and failed to find the container existence and when the program aborted (and killed the docker run process), the daemon promptly published the container. There in lies the problem. It looks as if the daemon lost an event that this container was created and when the "docker run" was killed that made the daemon check the status and find the container was created and hence published its state. This is definitely a container leak. The daemon should not have put out this container at all.

All these 127 containers should not have appeared after program ended with the daemon telling it none of them existed. Even if the daemon had got the container created, it should have thrown it away. But the worrying fact is did the daemon miss the event under load that the container was ready. Because that then signals a performance issue as well and not only correctness issue.

I have shortened the timeout in program and in that case I have seen for example maybe 300 running, 120 created after program termination and 80 not reported at all (guessing for those 80 daemon did not do any processing at all). 

I can also confirm tha docker daemon log did not have the POST request to start the containers for any of these 127 containers.
## So can I get 500 containers?

Yes, if I either have a larger timeout in the program, all 500 shows up running. Or If I spin in batches of 100, I reach 500. But the above behavior is concerning.

**Comment by mlaventure** [comment]
Ok, so I've come around to play a bit and this is my analysis: I do not think it is a bug per se but maybe more of performance issue.

When the script timeout is reached, it kills the call to `docker run` before it had the time to send the `run` api call (a `docker run` is made of 2 api calls `create` and `run`). However, by that time the `create` container request has already been received by the daemon and is in the process of being processed.

This explains why those containers appears as stopped after the script has ended. It is just the daemon catching up on all the queued up `create` request. The inspect do show that their status is `created` and not `stopped` as expected.

In a nutshell, given enough time all the containers would correctly show up and will be running.

Using a lighter container (e.g. busybox top) the issue won't appear until the timeout is significantly lowered.

**Comment by mbdas** [comment]
I disagree this being only performance issue. This is a issue of correctness. The daemon should not expose the container after docker run killed if it didn't so far publish it till the timeout. This is container leaking. This is different from container got published in create state but not move to run state. If daemon kept the creation state in memory where inspect and all others failed to find it, daemon should throw away the container. Who is going to clean up these containers?

**Comment by mbdas** [comment]
I disagree this being only performance issue. This is a issue of correctness. The daemon should not expose the container after docker run killed if it didn't so far publish it till the timeout. This is container leaking. This is different from container got published in create state but not move to run state. If daemon kept the creation state in memory where inspect and all others failed to find it, daemon should throw away the container. Who is going to clean up these containers?

**Comment by mlaventure** [comment]
@mbdas , Just trying to see if I got your point correctly.

You mean that if I write a client that send a `create` request to the daemon and the client exits/dies before  the daemon has processed the request fully, the resulting artifact (i.e. the container) should be automatically deleted by the daemon?

**Comment by mbdas** [comment]
Why not? Because the program kept polling the daemon in this case and daemon repeatedly said no such container. And then when program exited it published it immediately. That is wrong. And I do believe there is a performance issue because why couldn't the daemon recognize it till timeout and then immediately recognize it after termination. What was stuck?

**Comment by mbdas** [comment]
40 CPU, 157 gig memory , no other load, just 500 containers and over 3 and half mins of polling in this case. So we should look at performance but correctness is separate as mentioned above.

**Comment by mlaventure** [comment]
@mbdas It is not that the daemon didn't recognize the container but that the container hasn't finished being created when the inspect requests are being serviced.

The fact that they appear afterwards, it probably a race with the common lock(s) surrounding both `inspect` and `create` operations, and it seems that the inspect operation is getting it a lot more often that the `create` one.

**Comment by mbdas** [comment]
So can this be addressed? In actual mesos world the inspect is running every 500ms. I made it 1 sec in the simulated example.

**Comment by mlaventure** [comment]
Yes, it should be addressed, hence why the issue is still open :)

**Comment by mlaventure** [comment]
@mbdas btw, you should consider using `docker events` instead of polling via `inspect`

**Comment by mbdas** [comment]
As you see from my original comments, I had to construct the example from how mesos support for docker is built in. I am on the consumer side for both docker and mesos.

**Comment by unclejack** [comment]
The `docker info` output is telling us what's wrong:
```
Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
```

The point is that it takes a while to create the container itself. This can take a long time because devicemapper is being used on loopback mounted devices. This can be one of the reasons you're getting really bad performance.

Please try to switch to a setup with devicemapper on a real block device and let us know if that improved performance for you.

**Comment by mbdas** [comment]
I don't suspect that this is a big issue because I do feel based on earlier comments that internal docker lock contention in how the example code (simulating mesos native integration) has been set up is slowing things down. Just firing docker run (without any inspects) can make it go faster even in current settings. In our internal setup we use aufs. But point noted that for device mapper proper usage that point should be addressed. This was an exercise on a default largest box in AWS in context of a test mesos cluster. The correctness issue of container existence needs to be addressed that is orthogonal to the performance. So, if for some reasons things are slow and containers cannot be recognized through inspect and docker run dies, no need to publish that container.

**Comment by LK4D4** [comment]
@cpuguy83 @mlaventure 
Dunno how mesos works now, but not sure that anyone should rely on inspect for such purposes :/

**Comment by thaJeztah** [comment]
Let me close this ticket for now, as it looks like it went stale.
